nanofi/src/coap/c2protocol.c

```diff
@@ -0,0 +1,403 @@
+        cbor_item_t * pmap = cbor_new_indefinite_map();
+        properties_t * el, *tmp;
+        HASH_ITER(hh, value.v_props, el, tmp) {
+            cbor_map_add(pmap,
+                         (struct cbor_pair){
+                            .key = cbor_move(cbor_build_string(el->key)),
+                            .value = cbor_move(cbor_build_string(el->value))
+                         });
+        }
+        cbor_array_push(item, pmap);
+        break;
+    }
+    default:
+        break;
+    }
+}
+
+void build_cbor_map(cbor_item_t * cbor_map, uint16_t key, value_t value) {
+    if (!cbor_map || !cbor_isa_map(cbor_map)) return;
+
+    switch (value.val_type) {
+    case UINT8_TYPE: {
+        cbor_map_add(cbor_map,
+                     (struct cbor_pair){
+                         .key = cbor_move(cbor_build_uint16(key)),
+                         .value = cbor_move(cbor_build_uint8(value.v_uint8))});
+        break;
+    }
+    case UINT16_TYPE: {
+        cbor_map_add(cbor_map,
+                     (struct cbor_pair){
+                         .key = cbor_move(cbor_build_uint16(key)),
+                         .value = cbor_move(cbor_build_uint16(value.v_uint16))});
+        break;
+    }
+    case UINT32_TYPE: {
+        cbor_map_add(cbor_map,
+                     (struct cbor_pair){
+                         .key = cbor_move(cbor_build_uint16(key)),
+                         .value = cbor_move(cbor_build_uint32(value.v_uint32))});
+        break;
+    }
+    case UINT64_TYPE: {
+        cbor_map_add(cbor_map,
+                     (struct cbor_pair){
+                         .key = cbor_move(cbor_build_uint16(key)),
+                         .value = cbor_move(cbor_build_uint64(value.v_uint64))});
+        break;
+    }
+    case STRING_TYPE: {
+        cbor_map_add(cbor_map,
+                     (struct cbor_pair){
+                         .key = cbor_move(cbor_build_uint16(key)),
+                         .value = cbor_move(cbor_build_string(value.v_str))});
+        break;
+    }
+    case HASH_TYPE: {
+        cbor_item_t * map_item = cbor_new_indefinite_map();
+        struct c2_payload_map * el, *tmp;
+        HASH_ITER(hh, value.v_map, el, tmp) {
+            build_cbor_map(map_item, el->key, el->value);
+        }
+        cbor_map_add(cbor_map,
+                     (struct cbor_pair){
+                         .key = cbor_move(cbor_build_uint16(key)),
+                         .value = cbor_move(map_item)
+                     });
+        break;
+    }
+    case LIST_TYPE: {
+        cbor_item_t * list_item = cbor_new_indefinite_array();
+        c2_payload_list_t * el;
+        LL_FOREACH(value.v_maplist, el) {
+            build_cbor_list(list_item, el->value);
+        }
+        cbor_map_add(cbor_map,
+                     (struct cbor_pair){
+                         .key = cbor_move(cbor_build_uint16(key)),
+                         .value = cbor_move(list_item)});
+        break;
+    }
+    case PROP_TYPE: {
```

Comment:
All keys in the C2 protocol are unsigned ints except the keys of properties which happen to be strings. Use semantic tagging here to identify properties (string key and string value pairs). While decoding a cbor serialized data, tag will help determine if the incoming payload is a regular cbor map or a property.